File: src/stores/mcp.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { invoke } from "@tauri-apps/api/core";

export interface McpServerInfo {
  name: string;
  transport: string;
  connected: boolean;
  tools: number;
  tools_registered: number;
  tool_names: string[];
  tool_names_registered: string[];
  error: string | null;
  command: string | null;
  args: string[] | null;
  url: string | null;
  env: Record<string, string> | null;
  headers: Record<string, string> | null;
  tools_config: any;
  prompts: boolean | null;
  resources: boolean | null;
  enabled: boolean | null;
  raw_config: any;
  tool_details: Array<{ name: string; description: string }> | null;
}

export interface McpToolInfo {
  name: string;
  description: string;
  input_schema: any;
}

export const useMcpStore = defineStore("mcp", () => {
  const servers = ref<McpServerInfo[]>([]);
  const loading = ref(false);
  const error = ref("");
  const searchQuery = ref("");
  const toolsByServer = ref<Record<string, McpToolInfo[]>>({});

  const summary = computed(() => {
    let connected = 0;
    let totalTools = 0;
    for (const s of servers.value) {
      if (s.connected) connected++;
      totalTools += s.tools_registered;
    }
    return {
      total: servers.value.length,
      connected,
      disconnected: servers.value.length - connected,
      totalTools,
    };
  });

  const filteredServers = computed(() => {
    const q = searchQuery.value.trim().toLowerCase();
    if (!q) return servers.value;
    return servers.value.filter(
      (s) =>
        s.name.toLowerCase().includes(q) ||
        s.transport.includes(q) ||
        s.tool_names.some((n) => n.toLowerCase().includes(q))
    );
  });

  async function loadServers() {
    loading.value = true;
    error.value = "";
    try {
      const data: { servers: McpServerInfo[]; total_tools: number } =
        await invoke("mcp_list_servers");
      servers.value = data.servers ?? [];

      // Populate toolsByServer
      const next: Record<string, McpToolInfo[]> = {};
      for (const s of servers.value) {
        next[s.name] = (s.tool_details || []).map((t) => ({
          name: t.name,
          description: t.description || "",
          input_schema: null,
        }));
      }
      toolsByServer.value = next;
    } catch (e: any) {
      error.value = e?.message || "加载 MCP 服务器失败";
    } finally {
      loading.value = false;
    }
  }

  async function addServer(name: string, config: any) {
    const res: { ok: boolean; error?: string } = await invoke("mcp_add_server", {
      name,
      config,
    });
    if (!res.ok) throw new Error(res.error || "添加失败");
    await loadServers();
    return res;
  }

  async function updateServer(name: string, config: any) {
    const res: { ok: boolean; error?: string } = await invoke("mcp_update_server", {
      name,
      config,
    });
    if (!res.ok) throw new Error(res.error || "更新失败");
    await loadServers();
    return res;
  }

  async function removeServer(name: string) {
    const res: { ok: boolean; error?: string } = await invoke("mcp_remove_server", { name });
    if (!res.ok) throw new Error(res.error || "删除失败");
    const { [name]: _, ...rest } = toolsByServer.value;
    toolsByServer.value = rest;
    await loadServers();
    return res;
  }

  async function testServer(name: string) {
    const res: { ok: boolean; error?: string; tools?: string[] } = await invoke(
      "mcp_test_server",
      { name }
    );
    return res;
  }

  async function reload(server?: string) {
    const res: { ok: boolean; error?: string; message?: string } = await invoke("mcp_reload", {
      server: server || null,
    });
    if (server) {
      const { [server]: _, ...rest } = toolsByServer.value;
      toolsByServer.value = rest;
    } else {
      toolsByServer.value = {};
    }
    await loadServers();
    return res;
  }

  function statusClass(s: McpServerInfo): string {
    if (s.enabled === false) return "disabled";
    return s.connected ? "connected" : "disconnected";
  }

  function statusLabel(s: McpServerInfo): string {
    if (s.enabled === false) return "已禁用";
    return s.connected ? "已连接" : "未连接";
  }

  return {
    servers,
    loading,
    error,
    searchQuery,
    toolsByServer,
    summary,
    filteredServers,
    loadServers,
    addServer,
    updateServer,
    removeServer,
    testServer,
    reload,
    statusClass,
    statusLabel,
  };
});

File: src-tauri/src/mcp.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ═══════════════════════════════════════════
// Types — mirrors Hermes Studio MCP API
// ═══════════════════════════════════════════

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct McpServerInfo {
    pub name: String,
    pub transport: String,
    pub connected: bool,
    #[serde(default)]
    pub tools: u32,
    #[serde(default)]
    pub tools_registered: u32,
    #[serde(default)]
    pub tool_names: Vec<String>,
    #[serde(default)]
    pub tool_names_registered: Vec<String>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub command: Option<String>,
    #[serde(default)]
    pub args: Option<Vec<String>>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub env: Option<HashMap<String, String>>,
    #[serde(default)]
    pub headers: Option<HashMap<String, String>>,
    #[serde(default)]
    pub tools_config: Option<serde_json::Value>,
    #[serde(default)]
    pub prompts: Option<bool>,
    #[serde(default)]
    pub resources: Option<bool>,
    #[serde(default)]
    pub enabled: Option<bool>,
    #[serde(default)]
    pub raw_config: Option<serde_json::Value>,
    #[serde(default)]
    pub tool_details: Option<Vec<McpToolInfo>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct McpToolInfo {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub input_schema: Option<serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct McpServersResponse {
    pub servers: Vec<McpServerInfo>,
    #[serde(default)]
    pub total_tools: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct McpToolsResponse {
    #[serde(default)]
    pub ok: bool,
    #[serde(default)]
    pub results: Vec<McpToolResult>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct McpToolResult {
    pub server: String,
    pub tools: Vec<McpToolInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct McpActionResponse {
    pub ok: bool,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub tools: Option<Vec<String>>,
}

// ═══════════════════════════════════════════
// Hermes Studio API helpers
// ═══════════════════════════════════════════

fn hermes_base_url() -> String {
    std::env::var("HERMES_STUDIO_URL")
        .unwrap_or_else(|_| "http://localhost:8648".to_string())
}

fn hermes_client() -> reqwest::blocking::Client {
    reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(15))
        .build()
        .unwrap_or_default()
}

// ═══════════════════════════════════════════
// Tauri Commands
// ═══════════════════════════════════════════

#[tauri::command]
pub async fn mcp_list_servers() -> Result<McpServersResponse, String> {
    let url = format!("{}/api/hermes/mcp/servers", hermes_base_url());
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let data: McpServersResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_list_tools(server: Option<String>) -> Result<McpToolsResponse, String> {
    let mut url = format!("{}/api/hermes/mcp/tools", hermes_base_url());
    if let Some(s) = &server {
        url.push_str(&format!("?server={}", s));
    }
    let client = hermes_client();
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let data: McpToolsResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_add_server(name: String, config: serde_json::Value) -> Result<McpActionResponse, String> {
    let url = format!("{}/api/hermes/mcp/servers", hermes_base_url());
    let client = hermes_client();
    let body = serde_json::json!({ "name": name, "config": config });
    let resp = client.post(&url).json(&body).send().map_err(|e| e.to_string())?;
    let data: McpActionResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_update_server(name: String, config: serde_json::Value) -> Result<McpActionResponse, String> {
    let url = format!(
        "{}/api/hermes/mcp/servers/{}",
        hermes_base_url(),
        name
    );
    let client = hermes_client();
    let resp = client
        .patch(&url)
        .json(&config)
        .send()
        .map_err(|e| e.to_string())?;
    let data: McpActionResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_remove_server(name: String) -> Result<McpActionResponse, String> {
    let url = format!(
        "{}/api/hermes/mcp/servers/{}",
        hermes_base_url(),
        name
    );
    let client = hermes_client();
    let resp = client.delete(&url).send().map_err(|e| e.to_string())?;
    let data: McpActionResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_test_server(name: String) -> Result<McpActionResponse, String> {
    let url = format!(
        "{}/api/hermes/mcp/servers/{}/test",
        hermes_base_url(),
        name
    );
    let client = hermes_client();
    let resp = client.post(&url).send().map_err(|e| e.to_string())?;
    let data: McpActionResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn mcp_reload(server: Option<String>) -> Result<McpActionResponse, String> {
    let mut url = format!("{}/api/hermes/mcp/reload", hermes_base_url());
    if let Some(s) = &server {
        url.push_str(&format!("?server={}", s));
    }
    let client = hermes_client();
    let resp = client.post(&url).send().map_err(|e| e.to_string())?;
    let data: McpActionResponse = resp.json().map_err(|e| e.to_string())?;
    Ok(data)
}
